lib/memory.c

```diff
@@ -38,7 +38,12 @@ static inline void mt_count_alloc(struct memtype *mt, size_t size, void *ptr)
 {
 	size_t oldsize;
 
-	atomic_fetch_add_explicit(&mt->n_alloc, 1, memory_order_relaxed);
+	oldsize = 1 + atomic_fetch_add_explicit(&mt->n_alloc, 1,
+						memory_order_relaxed);
+
+	if (oldsize > atomic_load_explicit(&mt->n_max, memory_order_relaxed))
```

Comment:
we've got some code in lib/thread.c that's using `atomic_compare_exchange_weak_explicit()` in a loop to accomplish this sort of thing - maybe that pattern could be reused?